### DISCUSSION TITLE
Wake-On-Lan Support

### DISCUSSION THREAD

**Discussion by marvinlehmann** [farag2/Sophia-Script-for-Windows#344] [Ideas]
Hi,
I think it would be nice to have a function for enabling/disabling wake-on-lan. =)

# Settings for Wake-On-Lan
_0. Enabling it in BIOS/UEFI_

## 1. Disabling **fast startup** and/or hibernation
**Turning off fast startup:** 
```powershell
New-ItemProperty "HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Power" -Name HiberbootEnabled -PropertyType DWord -Value 0 -Force
````

**Disabling hibernation**, which disables the former too (?): https://github.com/farag2/Sophia-Script-for-Windows/blob/cb9af2a839c1a067cff9f5a18d24e2ea421d9432/Sophia%20Script/Sophia%20Script%20for%20Windows%2010%20PowerShell%207/Module/Sophia.psm1#L4003-L4055

## 2. **NIC Power Management**

**2.1: Checking "Allow the computer to turn off this device to save power"** ✅
https://github.com/farag2/Sophia-Script-for-Windows/blob/cb9af2a839c1a067cff9f5a18d24e2ea421d9432/Sophia%20Script/Sophia%20Script%20for%20Windows%2010%20PowerShell%207/Module/Sophia.psm1#L5572-L5634

**2.2: Checking "Allow this device to wake the computer"** ✅
```powershell
$PnPDeviceID = Get-NetAdapter -Physical | Where-Object { $_.PhysicalMediaType -eq "802.3" }
  | Select-Object -First 1 -ExpandProperty PnPDeviceID # only getting first one here for illustration...

Get-CimInstance -ClassName MSPower_DeviceWakeEnable -Namespace root/wmi
  | Where-Object { $_.InstanceName -Match [regex]::escape($PnPDeviceID) }
    | Set-CimInstance -Property @{Enable=$True}
```

**2.3: Checking "Only allow a magic packet to wake the computer"** ✅
```powershell
$PnPDeviceID = Get-NetAdapter -Physical | Where-Object { $_.PhysicalMediaType -eq "802.3" }
  | Select-Object -First 1 -ExpandProperty PnPDeviceID # only getting first one here for illustration...

Get-CimInstance -ClassName MSNdis_DeviceWakeOnMagicPacketOnly -Namespace root/wmi
  | Where-Object { $_.InstanceName -Match [regex]::escape($PnPDeviceID) }
    | Set-CimInstance -Property @{EnableWakeOnMagicPacketOnly=$true}
```
## 3. **Advanced NetAdapter Properties**
- **EnablePME** (Intel also calls it "Wake on Magic Packet from power off state"): Enabled
- **Wake on Magic Packet**: Enabled 
- **Wake on Pattern Match**: Enabled
- **Energy Efficient Ethernet**: Disabled? (mentioned in some instructions e.g. by DELL - for me it didn't matter)

````powershell
$Adapter | Set-NetAdapterPowerManagement -WakeOnMagicPacket Enabled
$Adapter | Set-NetAdapterPowerManagement -WakeOnPattern Enabled
$Adapter | Set-NetAdapterAdvancedProperty -RegistryKeyword EnablePME -RegistryValue 1
# wasn't needed for me:
$Adapter | Set-NetAdapterAdvancedProperty -RegistryKeyword EEELinkAdvertisement -RegistryValue 0 
````

**Comment by farag2** [comment]
🤔

**Reply by mrmattipants** [reply]
Great minds think alike ;)

**Reply by farag2** [reply]
😹
